lib/core/include/authentication_plugin_framework.hpp

```diff
@@ -0,0 +1,155 @@
+#ifndef IRODS_AUTHENTICATION_PLUGIN_FRAMEWORK_HPP
+#define IRODS_AUTHENTICATION_PLUGIN_FRAMEWORK_HPP
+
+#include "authCheck.h"
+#include "authPluginRequest.h"
+#include "authRequest.h"
+#include "authResponse.h"
+#include "authenticate.h"
+#include "irods_auth_constants.hpp"
+#include "irods_auth_factory.hpp"
+#include "irods_auth_manager.hpp"
+#include "irods_auth_plugin.hpp"
+#include "irods_kvp_string_parser.hpp"
+#include "irods_native_auth_object.hpp"
+#include "irods_stacktrace.hpp"
+#include "miscServerFunct.hpp"
+#include "msParam.h"
+#include "rcConnect.h"
+#include "rodsDef.h"
+
+#ifdef RODS_SERVER
+#include "rsAuthCheck.hpp"
+#include "rsAuthRequest.hpp"
+#endif // RODS_SERVER
+
+#include <boost/any.hpp>
+#include <fmt/format.h>
+#include <nlohmann/json.hpp>
+
+#include <openssl/md5.h>
+
+#include <termios.h>
+#include <unistd.h>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+using json = nlohmann::json;
+
+namespace irods::experimental
+{
+    namespace auth
+    {
+        static const std::string flow_complete{"authentication_flow_complete"};
+
+        static const std::string next_operation{"next_operation"};
+    }
+
+    class authentication_base : public irods::plugin_base
+    {
+    public:
+        #define OPERATION(C, F) std::function<json(C&, const json&)>([&](C& c, const json& j) -> json {return F(c, j);})
+
+        authentication_base()
+            : plugin_base{"authentication_framework_plugin", "empty_context_string"}
+        {
+            add_operation(AUTH_CLIENT_START, OPERATION(rcComm_t, auth_client_start));
+        } // ctor
+
+        virtual json auth_client_start(rcComm_t& comm, const json& req) = 0;
+
+        template<typename RxComm>
+        void add_operation(const std::string& n, std::function<json(RxComm&, const json&)> f)
+        {
+            if(operations_.find(n) != operations_.end()) {
+                THROW(SYS_INTERNAL_ERR, fmt::format("operation already exists [{}]", n));
+            }
+
+            if(n.empty()) {
```

Comment:
the if-blocks (including their bodies) on lines 65 and 69 seem like they should be swapped.